# IOS and Android apps analysis

For this project, we'll pretend we're working as data analysts for a company that builds Android and iOS mobile apps. We make our apps available on Google Play and the App Store.

Our apps are **free to dowload and install**, our main source of revenue consists of **in-app ads**. Our business is therefore linked to the number of downloads of our applications, and to the engagement of our users. 

Our goal for this project is to analyze data to **help our developers understand what type of apps are likely to attract more users**.

We'll start by opening and exploring these two data sets

First one : containing data about approximately 10,000 Android apps from Google Play

Second one : containing data about approximately 7,000 iOS apps from the App Store

In [1]:
#Opening Google Play Dataset

opened_file = open('googleplaystore.csv') 
from csv import reader
read_file = reader(opened_file) 
googleplay_data = list(read_file)
OS = googleplay_data[1:]
OS_header = googleplay_data[0]

#Opening AppleStore dataset

opened_file = open('AppleStore.csv') 
from csv import reader
read_file = reader(opened_file) 
applestore_data = list(read_file)
IOS = applestore_data[1:]
IOS_header = applestore_data[0]

We will use a function which has been defined in the exercise statement. It allows us to have an overview of our dataset and to identify the data that we can use.

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
print(OS_header)
explore_data(OS,0,5)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Every

In [4]:
print(IOS_header)
explore_data(IOS,0,5)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']




Before analysing the data, it must be cleaned up.We have two main tasks to carry out before analysing the data: 

- Remove non-English application data from the analysis
- Remove paid applications from the analysis, as our company only develops free applications. 

The comments in the Google Play database indicate an error on a specific row in the database.

In [5]:
print(OS[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [6]:
print('Before removing :',len(OS))
del OS[10472]
print('After removing :',len(OS))

Before removing : 10841
After removing : 10840


The comments also tell us about the presence of duplicates in the Google Play database. In the statement we were given a code to fix the duplicates (loop on the apps in OS, then if the app name already exists then it is a duplicate)

In [7]:
duplicate_apps =[]
unique_apps = []

for app in OS : 
    name = app[0]
    
    if name in unique_apps : 
        duplicate_apps.append(name)
    else :
        unique_apps.append(name)

print(len(duplicate_apps))

1181


The Instagram application comes up 4 times, and we notice a difference in the number of reviews. We will use this information to keep the most recent rows of duplicate applications.

In [8]:
for app in OS : 
    name = app[0]
    if name == 'Instagram':
        print(app)
    

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [9]:
print('Expected lenght :', len(OS)-len(duplicate_apps))

Expected lenght : 9659


In [10]:
reviews_max = {}

for app in OS : 
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews : 
        reviews_max[name] = n_reviews
    elif name not in reviews_max : 
        reviews_max[name] = n_reviews

print(len(reviews_max))

9659


In [11]:
android_clean = []
already_added = []

for row in OS : 
    name = row[0]
    n_reviews = float(row[3])
    
    if n_reviews == reviews_max[name] and name not in already_added : 
        android_clean.append(row)
        already_added.append(name)
        
print(len(android_clean))



9659


We have the right number of applications in the android_clean list (9659 = Sum of applications - duplicates)

We are now going to check if there are duplicates in the IOS database, for that we are going to base ourselves on the id column

In [12]:
duplicate_IOSapps =[]
unique_IOSapps = []

for row in IOS : 
    id_column = row[0]
    
    if id_column in unique_IOSapps : 
        duplicate_IOSapps.append(id_column)
    else :
        unique_IOSapps.append(id_column)

print('Duplicate IOS App :',len(duplicate_IOSapps))
    
    

Duplicate IOS App : 0


There is no duplicate in IOS database

To remove non-English applications we will use the function ord() which returns a number for each letter (e.g. ord(a) returns 97 and ord(€) returns 8364). The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII (American Standard Code for Information Interchange) system

In [13]:
# Function that returns False if the name contains non-English characters, and true if no non-English characters are found
def is_english (string): 
    for character in string : 
        if ord(character) > 127:
            
            return False
       
    
    return True

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


Some application names are not considered English, this is because emojis and characters like ™ fall outside the ASCII range and have corresponding numbers over 127.

In [14]:
# Function that returns False if the name contains more than 3 non-English characters, and true if no non-English characters are found
def is_english (string): 
    number_non_english_chr = 0
    for character in string :
        if ord(character) > 127:
            number_non_english_chr += 1
        
        if number_non_english_chr > 3 :
            
            return False
       
    
    return True

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
True
True


In [15]:
English_OS = []

for row in android_clean : 
    name = row[0]
    if is_english(name) == True :
        English_OS.append(row)

print('Number of English App :',len(English_OS))

Number of English App : 9614


In [16]:
English_IOS = []

for row in IOS : 
    name = row[1]
    if is_english(name) == True :
        English_IOS.append(row)

print('Number of English App :',len(English_IOS))

Number of English App : 6183


Our data sets contain both free and non-free apps; we'll need to isolate only the free apps for our analysis.

- Price column in Android database : 8th 
- Price column in Apple database : 5th 

In [17]:
print(type(English_IOS[0][4]))
print(type(English_OS[0][7]))

<class 'str'>
<class 'str'>


In [18]:
Free_IOS = []
Non_free_IOS = []

for row in English_IOS : 
    price = row[4]
    
    if price =='0.0':
        Free_IOS.append(row)
    else :
        Non_free_IOS.append(row)

print(len(Free_IOS))   
print(len(Non_free_IOS))

explore_data(Free_IOS,0,5)

3222
2961
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']




In [19]:
Free_OS = []
Non_free_OS = []

for row in English_OS : 
    price = row[7]
    
    if price =='0':
        Free_OS.append(row)
    else :
        Non_free_OS.append(row)

print(len(Free_OS))   
print(len(Non_free_OS))

explore_data(Free_OS,0,5)

8864
750
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']




## Finding the most common apps by genre

According to the statement :

"To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

- Build a minimal Android version of the app, and add it to Google Play.
- If the app has a good response from users, we develop it further.
- If the app is profitable after six months, we build an iOS version of the app and add it to the App Store."

In [20]:
#We will us genre column
print(OS_header[1])
print(IOS_header[11])

Category
prime_genre


To do this we will create a % frequency dictionary and a second dictionary to classify the results and interpret them more easily

In [21]:
def freq_table(dataset,index) : 
    frequency = {}
    total = 0
    for rows in dataset : 
        row = rows[index]
        total +=1
        
        if row not in frequency : 
            frequency[row] = 1
        else :
            frequency[row] +=1
            
    frequency_percentage = {}
    for row in frequency : 
        percentage = (frequency[row]/total)*100
        frequency_percentage[row] = percentage
        
    return frequency_percentage

#Function that was given in the statement, using to sort by % of frequency

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        

In [22]:
display_table(Free_OS,1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

There is a strong interest among Android mobile users in the following genres for mobile app: 

- Family
- Game
- Tools
- Business

Among the most represented genres we find mainly "serious" genres around productivity, business sector or Finance, health... and not so much application designed for fun, at the opposite of IOS app frequency

In [23]:
display_table(Free_IOS,11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Applications registered as "Games" represent more than half of the free IOS app available. If we add Entertainement and Photo & Video (we can see it as entertainement too), the IOS application directories are mainly focused on entertainment (more thant 70% of free app available)

Bonus : There is an other column in Google Play dataset that we can use, but it's too granular (see below) for our first goal wich is to understand the global picture of app market.

In [24]:
display_table(Free_OS,9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For OS dataset, we have this information in the Installs column (column 6) but we don't have this information on IOS dataset. For IOS analysis, we could use the number of user ratings as a proxy (column rating_count_tot, column 6)

In [25]:
prime_genreIOS = freq_table(Free_IOS,11)

for genre in prime_genreIOS : 
    total = 0
    len_genre = 0
    for row in Free_IOS : 
        genre_app = row[11]
        if genre_app == genre : 
            number_rating = float(row[5])
            total += number_rating
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


Navigation applications are the most downloaded applications in the AppleStore. But this performance is influenced by Waze and Google Maps, which have close to half a million user reviews together:

In [26]:
for app in Free_IOS : 
    if app[11] == 'Navigation' :
        print(app[1], app[5])

Waze - GPS Navigation, Maps & Real-time Traffic 345046
Google Maps - Navigation & Transit 154911
Geocaching® 12811
CoPilot GPS – Car Navigation & Offline Maps 3582
ImmobilienScout24: Real Estate Search in Germany 187
Railway Route Search 5


In [28]:
display_table(Free_OS,5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


We can see than most value gives information on the minimum number of downloads. For instance '+100.00' etc but we don't know if it's 100 000 or 200 000. We don't need precise data so we will say that +100 000 = 100 000 installs. 

First we need to replace '+' and ',' from installs data if we want this data to be considered as float

In [39]:
category_OS = freq_table(Free_OS,1)

for category in category_OS : 
    total = 0
    len_category = 0
    for row in Free_OS : 
        category_app = row[1]
        if category_app == category :
            installs_os = row[5]
            installs_os = installs_os.replace('+','')
            installs_os = installs_os.replace(',','')
            installs_os = float(installs_os)
            total += installs_os
            len_category += 1
            
    avg_rating = total/len_category
    avg_rating = round(avg_rating,0)
    print(category,' : ', avg_rating)


ART_AND_DESIGN  :  1986335.0
AUTO_AND_VEHICLES  :  647318.0
BEAUTY  :  513152.0
BOOKS_AND_REFERENCE  :  8767812.0
BUSINESS  :  1712290.0
COMICS  :  817657.0
COMMUNICATION  :  38456119.0
DATING  :  854029.0
EDUCATION  :  1833495.0
ENTERTAINMENT  :  11640706.0
EVENTS  :  253542.0
FINANCE  :  1387692.0
FOOD_AND_DRINK  :  1924898.0
HEALTH_AND_FITNESS  :  4188822.0
HOUSE_AND_HOME  :  1331541.0
LIBRARIES_AND_DEMO  :  638504.0
LIFESTYLE  :  1437816.0
GAME  :  15588016.0
FAMILY  :  3695642.0
MEDICAL  :  120551.0
SOCIAL  :  23253652.0
SHOPPING  :  7036877.0
PHOTOGRAPHY  :  17840110.0
SPORTS  :  3638640.0
TRAVEL_AND_LOCAL  :  13984078.0
TOOLS  :  10801391.0
PERSONALIZATION  :  5201483.0
PRODUCTIVITY  :  16787331.0
PARENTING  :  542604.0
WEATHER  :  5074486.0
VIDEO_PLAYERS  :  24727872.0
NEWS_AND_MAGAZINES  :  9549178.0
MAPS_AND_NAVIGATION  :  4056942.0


It seems like Communication app have the most installs (around 38,5M) 

In [41]:
for app in Free_OS : 
    if app[1] == 'COMMUNICATION' :
        print(app[0], app[5])

WhatsApp Messenger 1,000,000,000+
Messenger for SMS 10,000,000+
My Tele2 5,000,000+
imo beta free calls and text 100,000,000+
Contacts 50,000,000+
Call Free – Free Call 5,000,000+
Web Browser & Explorer 5,000,000+
Browser 4G 10,000,000+
MegaFon Dashboard 10,000,000+
ZenUI Dialer & Contacts 10,000,000+
Cricket Visual Voicemail 10,000,000+
TracFone My Account 1,000,000+
Xperia Link™ 10,000,000+
TouchPal Keyboard - Fun Emoji & Android Keyboard 10,000,000+
Skype Lite - Free Video Call & Chat 5,000,000+
My magenta 1,000,000+
Android Messages 100,000,000+
Google Duo - High Quality Video Calls 500,000,000+
Seznam.cz 1,000,000+
Antillean Gold Telegram (original version) 100,000+
AT&T Visual Voicemail 10,000,000+
GMX Mail 10,000,000+
Omlet Chat 10,000,000+
My Vodacom SA 5,000,000+
Microsoft Edge 5,000,000+
Messenger – Text and Video Chat for Free 1,000,000,000+
imo free video calls and chat 500,000,000+
Calls & Text by Mo+ 5,000,000+
free video calls and chat 50,000,000+
Skype - free IM & video

In [49]:
under_100_m = []

for app in Free_OS:
        installs_os = row[5]
        installs_os = installs_os.replace('+','')
        installs_os = installs_os.replace(',','')
        installs_os = float(installs_os)
        if (app[1] == 'COMMUNICATION') and installs_os < 100000000:
            under_100_m.append(installs_os)

sum(under_100_m) / len(under_100_m)

[10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 10000000.0, 100

10000000.0